go-controller/pkg/node/iptables/iptables_manager.go

```diff
@@ -0,0 +1,371 @@
+		return nil
+	}
+	c.chainRules[ch] = make([]RuleArg, 0)
+	return c.reconcile()
+}
+
+func (c *Controller) DeleteRule(table iptables.Table, chain iptables.Chain, proto iptables.Protocol, ruleArg RuleArg) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	var err error
+	if proto == iptables.ProtocolIPv4 {
+		err = execIPTablesWithRetry(func() error {
+			err = c.iptV4.DeleteRule(table, chain, ruleArg.Args...)
+			if err != nil {
+				klog.Errorf("IPTables Manager: err while deleting rule %v chain %v", ruleArg.Args, err)
+			}
+			return err
+		})
+		if err != nil {
+			return fmt.Errorf("failed to delete IPv4 rule %v on table %s and chain %s: %v", ruleArg.Args, table, chain, err)
+		}
+	} else {
+		err = execIPTablesWithRetry(func() error {
+			err = c.iptV6.DeleteRule(table, chain, ruleArg.Args...)
+			if err != nil {
+				klog.Errorf("IPTables Manager: err while deleting rule %v chain %v", ruleArg.Args, err)
+			}
+			return err
+		})
+		if err != nil {
+			return fmt.Errorf("failed to IPv6 delete rule %v on table %s and chain %s: %v", ruleArg.Args, table, chain, err)
+		}
+	}
+	ch := Chain{
+		Table: table,
+		Chain: chain,
+		Proto: proto,
+	}
+	temp := make([]RuleArg, 0)
+	for _, existingRuleArg := range c.chainRules[ch] {
+		if !existingRuleArg.equal(ruleArg) {
+			temp = append(temp, existingRuleArg)
+		}
+	}
+	c.chainRules[ch] = temp
+	return nil
+}
+
+// EnsureRules ensures only the subset of rules specified in the function description will be present on the chain
+func (c *Controller) EnsureRules(table iptables.Table, chain iptables.Chain, proto iptables.Protocol, newRuleArgs []RuleArg) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	newChain := Chain{
+		Table: table,
+		Chain: chain,
+		Proto: proto,
+	}
+	wantedRuleArgs, alreadyExists := c.chainRules[newChain]
+	if !alreadyExists {
+		c.chainRules[newChain] = newRuleArgs
+	} else {
+		for _, newRuleArg := range newRuleArgs {
+			alreadyExists = false
+
+			for _, existingRuleArg := range wantedRuleArgs {
+				if existingRuleArg.equal(newRuleArg) {
+					alreadyExists = true
+					break
+				}
+			}
+			if !alreadyExists {
+				c.chainRules[newChain] = append(c.chainRules[newChain], newRuleArg)
+			}
+		}
+	}
+	return c.reconcile()
+}
+
+// GetIPv4ChainRuleArgs returns IPv4 RuleArgs
+func (c *Controller) GetIPv4ChainRuleArgs(table iptables.Table, chain iptables.Chain) ([]RuleArg, error) {
+	return getChainRuleArgs(c.iptV4, table, chain)
+}
+
+// GetIPv6ChainRuleArgs returns IPv6 RuleArgs
+func (c *Controller) GetIPv6ChainRuleArgs(table iptables.Table, chain iptables.Chain) ([]RuleArg, error) {
+	return getChainRuleArgs(c.iptV6, table, chain)
+}
+
+func getChainRuleArgs(ipt iptables.Interface, table iptables.Table, chain iptables.Chain) ([]RuleArg, error) {
+	buf := bytes.NewBuffer(nil)
+	err := execIPTablesWithRetry(func() error {
+		if err := ipt.SaveInto(table, buf); err != nil {
+			return fmt.Errorf("failed to retrieve iptables table %s: %v", table, err)
```

Comment:
hi @martinkennelly ! I'm getting into trouble when starting latest ovnk8 with is on latest upstream here:

`kind.sh -ic -ds`

```
F0221 15:19:53.017986   11904 ovnkube.go:136] failed to start node network manager: failed to start default node network controller: failed to run egress IP controller: failed to run EgressIP controller because repairing node failed: failed to list IPTable IPv4 rules: failed to retrieve iptables table nat: exit status 1
```

Is this a known issue?

```
❯ kn get all
NAME                                         READY   STATUS             RESTARTS        AGE
pod/ovnkube-control-plane-84b6447657-gzldk   1/1     Running            0               34m
pod/ovnkube-identity-65cbc8cbff-cv6gc        1/1     Running            0               34m
pod/ovnkube-node-bt8zz                       5/6     CrashLoopBackOff   6 (2m10s ago)   9m56s
pod/ovnkube-node-jh2rb                       5/6     CrashLoopBackOff   6 (2m19s ago)   9m56s
pod/ovnkube-node-jtv2l                       5/6     CrashLoopBackOff   6 (2m14s ago)   9m56s
pod/ovs-node-7f4cb                           1/1     Running            0               34m
pod/ovs-node-h8k5m                           1/1     Running            0               34m
pod/ovs-node-x4xgp                           1/1     Running            0               34m

NAME                          DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR            AGE
daemonset.apps/ovnkube-node   3         3         0       3            0           kubernetes.io/os=linux   34m
daemonset.apps/ovs-node       3         3         3       3            3           kubernetes.io/os=linux   34m

NAME                                    READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/ovnkube-control-plane   1/1     1            1           34m
deployment.apps/ovnkube-identity        1/1     1            1           34m

NAME                                               DESIRED   CURRENT   READY   AGE
replicaset.apps/ovnkube-control-plane-84b6447657   1         1         1       34m
replicaset.apps/ovnkube-identity-65cbc8cbff        1         1         1       34m
```

Reply:
o/ @flavio-fernandes great to hear from you here!
The error comes from here: https://github.com/ovn-org/ovn-kubernetes/blob/2d988b178ec3569482c46af92f67eb9763562010/go-controller/pkg/node/iptables/iptables_manager.go#L214
So the `iptables-save` call exited with 1. I would try to modify the code a bit to get the [stderr](https://github.com/ovn-org/ovn-kubernetes/blob/1d5d3b1e2ce64989358182e728f72c51699dae3f/go-controller/vendor/k8s.io/kubernetes/pkg/util/iptables/iptables.go#L366) output:
```
  return fmt.Errorf("failed to retrieve iptables table %s, err: %v, stderr: %s", table, err, buf.String())
```
This should be more helpful. I think this error is likely related to the OS you are running on as I am not able to repro it myself.

Reply:
Flavio confirmed its due to an ipv6 nat kern module thats not loaded on his distro. I am aware of it. Ill fix it shortly.

Reply:
TY @martinkennelly and @kyrtapz !!
I was able to work around by explicitly loading the ip6 nat module.

```
# modprobe ip6table_nat
# lsmod | grep -i ip6table_nat
```